In [18]:
# generating two correlated predictor variables
set.seed(187)
x1 = rnorm(200, 10, 2)
x2 = 0.5*x1 + rnorm(200, 0, 4)
y = 0.7*x1 + 2.2*x2 + rnorm(200, 0, 4)
# fitting a multiple-regression model
m = lm(y~x1+x2)
coefs = summary(m)$coef

coefs

             Estimate Std. Error    t value     Pr(>|t|)
(Intercept) 0.4872150 1.34744782  0.3615836 7.180507e-01
x1          0.6417832 0.13245620  4.8452488 2.555991e-06
x2          2.1844639 0.06421607 34.0174053 2.101837e-84

In [19]:
summary(m)


Call:
lm(formula = y ~ x1 + x2)

Residuals:
    Min      1Q  Median      3Q     Max 
-8.4276 -2.7240 -0.0065  2.7041  9.7580 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)  0.48722    1.34745   0.362    0.718    
x1           0.64178    0.13246   4.845 2.56e-06 ***
x2           2.18446    0.06422  34.017  < 2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 3.618 on 197 degrees of freedom
Multiple R-squared:  0.8683,	Adjusted R-squared:  0.8669 
F-statistic: 649.3 on 2 and 197 DF,  p-value: < 2.2e-16


In [20]:
print(coefs[1,1])
print(coefs[2,1])
print(coefs[3,1])

# get predicted coefficients and multiply them by each predictor 
# (e.g. all x values for height, all x values for weight)
y_hat = coefs[1,1] + coefs[2,1]*x1 + coefs[3,1]*x2
var(y_hat)

[1] 0.487215
[1] 0.6417832
[1] 2.184464


[1] 85.4221

In [21]:
var(y_hat)/var(y)

[1] 0.8682827

In [22]:
# calculating variance explained by x1 
# (multiply coefficient by the mean of x2)
y_hat1 = coefs[1,1] + coefs[2,1]*x1 + coefs[3,1]*mean(x2)
var(y_hat1)

[1] 1.608668

In [23]:
var(y_hat1)/var(y)

[1] 0.01635149

In [24]:
# calculating variance explained by x2 
# (multiply coefficient by the mean of x1)
y_hat2 = coefs[1,1] + coefs[2,1]*mean(x1) + coefs[3,1]*x2
var(y_hat2)

[1] 79.29333

In [25]:
var(y_hat2)/var(y)

[1] 0.8059861

In [26]:
var(y_hat)

[1] 85.4221

In [27]:
var(y_hat1) + var(y_hat2)

[1] 80.902

In [28]:
# You need to add the 2*cov(y_hat1, y_hat2) in order to get the total variance
var(y_hat1) + var(y_hat2) + 2*cov(y_hat1, y_hat2)

[1] 85.4221

In [29]:
coefs[2:3,1]

       x1        x2 
0.6417832 2.1844639 

In [30]:
x1_z = (x1 - mean(x1))/sd(x1)
x2_z = (x2 - mean(x2))/sd(x2)
m = lm(y ~ x1_z + x2_z)
summary(m)


Call:
lm(formula = y ~ x1_z + x2_z)

Residuals:
    Min      1Q  Median      3Q     Max 
-8.4276 -2.7240 -0.0065  2.7041  9.7580 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)  19.4090     0.2558  75.866  < 2e-16 ***
x1_z          1.2683     0.2618   4.845 2.56e-06 ***
x2_z          8.9047     0.2618  34.017  < 2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 3.618 on 197 degrees of freedom
Multiple R-squared:  0.8683,	Adjusted R-squared:  0.8669 
F-statistic: 649.3 on 2 and 197 DF,  p-value: < 2.2e-16


In [31]:
x1_m = (x1 - mean(x1))/mean(x1)
x2_m = (x2 - mean(x2))/mean(x2)
summary(lm(y ~ x1_m + x2_m))


Call:
lm(formula = y ~ x1_m + x2_m)

Residuals:
    Min      1Q  Median      3Q     Max 
-8.4276 -2.7240 -0.0065  2.7041  9.7580 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)  19.4090     0.2558  75.866  < 2e-16 ***
x1_m          6.5254     1.3468   4.845 2.56e-06 ***
x2_m         12.3964     0.3644  34.017  < 2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 3.618 on 197 degrees of freedom
Multiple R-squared:  0.8683,	Adjusted R-squared:  0.8669 
F-statistic: 649.3 on 2 and 197 DF,  p-value: < 2.2e-16


The value of x1_m being 6.5254 means:
If x1 increases by 1%, y increases by 6.525 units, holding x2_m constant.

In [ ]:
m1 = lm(x1~x2)
r2 = summary(m1)$r.squared
1/(1-r2)

[1] 1.041714